lib/model/indexhandler.go

```diff
@@ -111,6 +111,42 @@ func newIndexHandler(conn protocol.Connection, downloads *deviceDownloadState, f
 	}
 }
 
+// waitCondLockedWithCtx waits for cond to signal, or for the context to finish.
+// If the context is cancelled, the cond will broadcast and an error will be returned.
+// cond must be locked when calling this function.
+func waitCondLockedWithCtx(cond *sync.Cond, ctx context.Context) error {
+	c := make(chan struct{}, 1)
```

Comment:
Whats the 1-buffering for?

Reply:
> Good catch, thanks!
> 
> Suggestion:
> Instead of creating a goroutine and channel on every `Wait`, you could have a single goroutine in `Serve` that broadcasts on `ctx.Done()` and check for `ctx.Done()` inside the for-loop after waiting.

Ah, good idea. will do!

> Whats the 1-buffering for?

Not really needed, you're right.